services/discovery/src/app.py

```diff
@@ -0,0 +1,227 @@
+            return assoc_in(world, ['coeffects', name], data)
+        return w
+    return d
+
+
+@coeffect('cloudtrail')
+def coeffects_cloudtrail(world):
+    response = ct.describe_trails()
+    return keyfilter(lambda x: x in {'trailList'}, response)
+
+
+@coeffect('s3')
+def coeffects_s3(world):
+    response = s3.list_buckets()
+    return keyfilter(lambda x: x in {'Buckets'}, response)
+
+
+@coeffect('cur')
+def coeffects_cur(world):
+    response = cur.describe_report_definitions()
+    return keyfilter(lambda x: x in {'ReportDefinitions'}, response)
+
+
+#####################
+#
+# Business Logic
+#
+#####################
+def audit(world):
+    trail_bucket = get_in(['coeffects', 'cloudtrail', 'trailList', 0, 'S3BucketName'], world)
+    local_buckets = {x['Name'] for x in get_in(['coeffects', 's3', 'Buckets'], world, [])}
+    output = {
+        'IsAuditAccount': trail_bucket in local_buckets,
+        'AuditCloudTrailBucketName': trail_bucket,
+    }
+    return update_in(world, ['output'], lambda x: merge(x or {}, output))
+
+
+def connected(world):
+    output = {
+        'IsConnectedAccount': True,
+    }
+    return update_in(world, ['output'], lambda x: merge(x or {}, output))
+
+
+def safe_check(schema, data):
+    try:
+        return schema(data)
+    except Exception:
+        return None
+
+
+def keep_valid(schema, xs):
+    return [
+        y for y in [safe_check(schema, x) for x in xs]
+        if y is not None
+    ]
+
+
+MINIMUM_CLOUDTRAIL = Schema({
+    "S3BucketName": str,
+    "SnsTopicName": str,
+    "SnsTopicARN": str,
+    "IsMultiRegionTrail": True,
+    "TrailARN": str,
+}, extra=ALLOW_EXTRA, required=True)
+
+
+IDEAL_CLOUDTRAIL = MINIMUM_CLOUDTRAIL.extend({
+    "IsOrganizationTrail": True,
+}, extra=ALLOW_EXTRA, required=True)
+
+
+def cloudtrail(world):
+    trails = get_in(['coeffects', 'cloudtrail', 'trailList'], world, [])
+    valid_trails = keep_valid(IDEAL_CLOUDTRAIL, trails) or keep_valid(MINIMUM_CLOUDTRAIL, trails)
+    trail_topic = valid_trails[0]['SnsTopicName'] if valid_trails else None
+    account_id = trail_topic.split(':')[4] if trail_topic else None
+    output = {
+        'IsCloudTrailAccount': account_id == event_account_id(world),
+        'CloudTrailSNSTopicName': trail_topic,
+    }
+    return update_in(world, ['output'], lambda x: merge(x or {}, output))
+
+
+MINIMUM_BILLING_REPORT = Schema({
+    'TimeUnit': 'HOURLY',
+    'Format': 'textORcsv',
+    'Compression': 'GZIP',
+    'AdditionalSchemaElements': ExactSequence(['RESOURCES']),
+    'S3Bucket': str,
+    'S3Prefix': str,
+    'S3Region': str,
+    'ReportVersioning': 'CREATE_NEW_REPORT',
+    'RefreshClosedReports': True,
+}, extra=ALLOW_EXTRA, required=True)
+
+
+def master_payer(world):
+    report_definitions = get_in(['coeffects', 'cur', 'ReportDefinitions'], world, [])
+    local_buckets = {x['Name'] for x in get_in(['coeffects', 's3', 'Buckets'], world, [])}
```

Comment:
local buckets?  As in "buckets in the same account?"

Reply:
yes